### ISSUE TITLE
[Question] In-place update of parameters in HMC/NUTS

### ISSUE THREAD

**Issue by jodablei** [pyro-ppl/pyro#1866]
Hi,

while doing some experiments I ran into some unexpected behaviour, so I ran the debugger and was checking some of the internal functions in HMC/NUTS. At some point, the function ```_single_step_verlet``` is called and is responsible for updating _inplace_ the parameters of interest in the model. However, the operations are not inplace, but replacing one tensor for another.

Consider the following simple example:
```python
import torch
a = torch.tensor(10.)
id_a = id(a)
a = a + 1
print(id(a) == id_a)  # False, not inplace op
id_a = id(a)
a += 1
print(id(a) == id_a)  # True, inplace op

```

The fix is relatively simple:
```python
def _single_step_verlet(z, r, potential_fn, inverse_mass_matrix, step_size, z_grads=None):
    r"""
    Single step velocity verlet that modifies the `z`, `r` dicts in place.
    """

    z_grads = _potential_grad(potential_fn, z)[0] if z_grads is None else z_grads

    for site_name in r:
        r[site_name] += 0.5 * step_size * (-z_grads[site_name])  # r(n+1/2)

    r_grads = _kinetic_grad(inverse_mass_matrix, r)
    for site_name in z:
        z[site_name] += step_size * r_grads[site_name]  # z(n+1)

    z_grads, potential_energy = _potential_grad(potential_fn, z)
    for site_name in r:
        r[site_name] += 0.5 * step_size * (-z_grads[site_name])  # r(n+1)

    return z, r, z_grads, potential_energy
```

**Comment by jodablei** [comment]
After closer inspection, the not in place operations might be needed after all. The dicts are not being deep copied in ```velocity_verlet``` and it seems the algorithm has been designed not to work with the original object parameters of a given model, but with copies that need to be set up every time the potential energy is evaluated. I believe it would be possible to modify the algo so as to make it in place, but probably not worth because it would require other changes, such as reverting the state of the dict ```z``` after any move that is not a sample. Any additional clarifications would be appreciated.

**Comment by neerajprad** [comment]
> At some point, the function _single_step_verlet is called and is responsible for updating inplace the parameters of interest in the model. 

The function does not intend to in-place  update the tensors representing the latent variables because the original tensors are required when evaluating (later) the energy difference, probability of acceptance for the proposal, and reinstated if the proposal is rejected. Could you add more details around what your model is, and how you are using HMC? I think that's the place where there is a mismatch between what you might be trying to do, and what is currently supported.

**Comment by fehiepsi** [comment]
Hi @jodablei, I think that modifying states in-place is not compatible to MCMC because we need to collect samples during the way. On the other hand, in NUTS, we need to maintain the values at some leaves of the tree.

**Comment by jodablei** [comment]
Yes, I eventually figured it out. At the time, I was not fully aware that the tensors in the argument dict ```z``` need to be set as the parameter objects of the model we are evaluating. Copying their values will not work, because then we won't be able to collect the gradients in HMC (we only have references to the tensors in ```z```). However, if the tensors in the dict were to be the original parameter objects of the model, replacing objects each time we evaluate the potential function would not be needed with in-place updates. I'm almost certain that it would be possible to make it work with in place updates, by caching a deep copy of ```z``` for each state we wish to revert back to, but again it's not really needed.

Unless you want to add more info on this subject, this can be closed. Thanks!

**Comment by neerajprad** [comment]
I am closing this issue; please feel free to ask any follow up questions on the [forum](https://forum.pyro.ai/).
